packages/core-bridge/ts/worker-tuner.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * @experimental
+ *
+ * A worker tuner allows the customization of the performance characteristics of workers by
+ * controlling how "slots" are handed out for different task types. In order to poll for and then
+ * run tasks, a slot must first be reserved by the {@link SlotSupplier} returned by the tuner.
+ */
+export interface WorkerTuner {
+  workflowTaskSlotSupplier: SlotSupplier;
+  activityTaskSlotSupplier: SlotSupplier;
+  localActivityTaskSlotSupplier: SlotSupplier;
+}
+
+/**
+ * @experimental
+ *
+ * Controls how slots are handed out for a specific task type.
+ *
+ * For now, only {@link ResourceBasedSlotOptions} and {@link FixedSizeSlotSupplier} are supported,
+ * but we may add support for custom tuners in the future.
+ */
+export type SlotSupplier = ResourceBasedSlotsForType | FixedSizeSlotSupplier;
+
+/**
+ * @experimental
+ *
+ * Options for a specific slot type within a {@link ResourceBasedSlotsForType}
+ */
+export interface ResourceBasedSlotOptions {
+  // Amount of slots that will be issued regardless of any other checks
+  minimumSlots: number;
+  // Maximum amount of slots permitted
+  maximumSlots: number;
+  // Minimum time we will wait (after passing the minimum slots number) between handing out new
+  // slots in milliseconds.
+  rampThrottleMs: number;
+}
+
+type ResourceBasedSlotsForType = ResourceBasedSlotOptions & {
+  tunerOptions: ResourceBasedTunerOptions;
+};
+
+/**
+ * @experimental
+ *
+ * Options for a {@link ResourceBasedTuner} to control target resource usage
+ */
+export interface ResourceBasedTunerOptions {
+  // A value between 0 and 1 that represents the target (system) memory usage. It's not recommended
+  // to set this higher than 0.8, since how much memory a workflow may use is not predictable, and
+  // you don't want to encounter OOM errors.
+  targetMemoryUsage: number;
```

Comment:
I'm ok with this description as long as resource-based tuners are experimental.

But before we reach GA, we'll need to better understand what is accounted as "memory usage" and "available memory" for the purpose of this tuner, and adjust the description accordingly.

Among other things:
- How does "memory usage" relates with Node's heap? e.g. `--max-old-space-size`?
- Is this memory usage/available memory of the process or the system? 
- What's the impact of Worker Threads?
- Are we sure that available memory is properly determined in containerized environments?

Reply:
> So nothing specific to node should really matter that much, all that matters is how the system usage is being reported.

Yeah… and that's why I'm concerned that this might actually lead to very unintuitive behaviors.

For context, Node (and I think that's partially true of all of our core-based SDKs languages) has its own notion of "heap", which gets allocated in large blocks, and then heap objects get allocated from that heap. That's the memory that is GC managed. The problem is that "system memory usage" is often not representative (i.e. not proportional to) of "heap usage", which I think could result in a few pathological cases.

For example, let's imagine that I run a worker with `targetMemoryUsage` set to 80%, on a system with 64 GB of RAM, without configuring Node's `--max-old-space-size` (or set it to a value <60%). IIRC, Node will by default limit its heap size to the smallest of (25% of available system memory or 4GB). With that config, system memory usage will never grow even near the 80% (4GB / 64GB, plus some overhead for non-heap stuff would put us at ~10%). Consequently, the `ResourceBasedAllocator` will very likely raise all `maxXxxSlot` parameters to their max values, but then the process dies from running out of heap…

I think the opposite direction is also possible. For various reasons, many GC implementations have tend to be reluctant to hand back heap blocks to the system. So it might be possible for memory usage to quickly grow to the target system memory usage, yet heap usage to be relatively low.

All I'm saying is that we'll need to play with this more to see how well that works. Of course, some of this may result in refinement to the engine itself, but we'll most certainly also need to clarify documentation…

Reply:
Yeah, that makes good sense. The usage determination interface is pluggable in Core, so we can implement different ways of measuring.